<a href="https://colab.research.google.com/github/iliza5522/QM22/blob/main/Copy_of_S2_2025_STUDENTS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Python and Data Science

**Overview of the session**

1.	Python for Data Analysis: Introduction to NumPy and Pandas
2.	Python for Data Visualisation: Using Matplotlib

## Quick recap

### Data types
numbers and operations


In [ ]:
x = 5
print(x)
print(x+1)

5
6


strings, use the `str()` function to turn variables into strings

In [ ]:
print("I have " + str(x) + " apples.")

lists are zero indexed

elements inside lists can be changed

In [ ]:
my_list = ['hi', 5]
my_list[0]

'hi'

In [ ]:
my_list[1] = 'm8'
print(my_list)

['hi', 'm8']


dictionaries contain key-value pairs

In [ ]:
my_icd10_dict = {
    "S93.1": "Dislocation of toe(s)",
    "S93.4": "Sprain and strain of ankle",
    "S93.5": "Sprain and strain of toe(s)"}

my_icd10_dict['S93.1']

'Dislocation of toe(s)'

In [ ]:
# show all keys of dictionary:
my_icd10_dict.keys()

dict_keys(['S93.1', 'S93.4', 'S93.5'])

### Conditional statements and loops
if, elif, else statements

In [ ]:
y = 2
if y == 0:
  print("Zero!")
elif y % 2 == 0:
  print(f"{y} is even")
else:
  print(str(y) + " is odd")

2 is even


for loop

In [ ]:
list(range(5))

In [ ]:
for i in range(5):
  if i == 0:
    print("Zero!")
  elif i % 2 == 0:
    print(f"{i} is even")
  else:
    print(str(i) + " is odd")

while loop

In [ ]:
i = 0
while i < 7:
  print("still counting, ", i)
  i += 1
print("finished counting")

### **Functions**

Before starting today's lecture, let's quickly sidetrack and cover what a **function** is. We've been using them throughout the course, but it's worth seeing how they're structured and how we can create our own functions.

A function takes the given inputs, and uses them to get to a desired output. They are used extensively in python to avoid repeating sections of code.

Below is a function called **_adder_** that takes in two numbers and returns the sum of the two numbers.

A function in Python is defined by using the **def** keyword, followed by a space and the name of the function.

Arguments are specified after the function name, inside the parentheses. You can add as many arguments as you want, separating them with a comma.

Arguments with an equal sign defines default values.

Our _adder_ function has two arguments (n1, n2). When the function is called, we pass along the first number we want to add and then the second number after the comma, which is then used inside the function to return the sum of the arguments.


In [ ]:
def adder(n1, n2):
  total = n1 + n2
  return total

sum_of_numbers = adder(2, 3)
print(sum_of_numbers)

# alternatively, we could've written the two lines above in a single line as shown below

print(adder(2, 3))

5
5


In [ ]:
def adder(n1, n2=1):
  '''
  This function adds 2 numbers together. This function adds 1 by default
  '''
  total = n1 + n2
  return total

print(adder(2)) # no n2 specified
print(adder(2,3)) # the function works as before

3
5


**Task: Tying together all the previous concepts** write a for loop to loop over the following list, printing every element as you go, and print the 4th element as the word four

`this_list = [0,1,8,4,4,4,7]`

**Extension 1** create a dictionary, and use it to be able to swap any digit to it's word counterpart. Try changing the 3rd or 6th element to see that your code is working.

**Extension 2** replace the for loop with a while loop

In [ ]:
# your answer
def conversion(this_list, conversion_dict):
  for i in range(len)


**Task:** Using the code you written earlier, write a function that takes any list of numbers and prints them, printing the word four at the 4th index.

Test your function with another list
`another_list = [-3,1,4,6,2,1,7]`


In [ ]:
#Exercise
this_list = [0,1,8,4,4,4,7]

for index, element in enumerate(this_list):
  if index == 3: # The 4th element is at index 3
    print("four")
  else:
    print(element)


0
1
8
four
4
4
7


In [ ]:
this_list = [0,1,8,4,4,4,7]
this_list[3]

english_words = {}
for position in range(len(this_list)):
  #do something
  print(n)
  if position == 3:
    print("four")
  else:
      print(this_list[position])

7
0
7
1
7
8
7
four
7
4
7
4
7
7


**Extension task** Extend your function by adding a second argument that defines the special index (previously 4th index) with the default value of 3. Your function should now print the index as a word at the appropriate point. Test different values to see your function in action.

In today's session, we will look to use two Python **libraries**, NumPy and Matplotlib, to help us analyse data. Python libraries contain pre-written functions that we can use without having to rewrite them from scratch.

## Introduction to NumPy and NumPy Arrays

While lists are very versatile data structures that can be used in Python, there are times when we need to analyse vast quantities of data and perform mathematical operations on them. We are going to use the NumPy ("Numerical Python") library to help us with this.

In order to access NumPy and its functionalities, we do the following:

In [ ]:
import numpy as np

To convert a list into a NumPy array, we simply do the following:     

In [ ]:
heights = [190, 160, 170, 180, 176]

heights_array = np.array(heights)

heights_array + 1

array([191, 161, 171, 181, 177])

We are now going to look at calculating the body mass index (BMI) for our group of people. To do this we will need to take their masses and divide by their height squared.

Notice that our array of heights is in centimetres. However, we need this array to be in metres. The beauty of numpy arrays is that we can perform mathematical operations in a single line of code. For example, to calculate the height of each member of the group in metres, we can simply divide the array by 100 and it will perform the calculation for each element of the array. For example:



In [ ]:
heights_metres = heights_array/100
print(heights_metres)

[1.9  1.6  1.7  1.8  1.76]


We can also perform operations between arrays. There are some caveats that depend on the shapes of the arrays, but these are beyond the scope of what will be discussed today. A basic example:

In [ ]:
distance_run_day1 = np.array([5, 10, 23, 11, 4])
distance_run_day2 = np.array([10, 20, 1, 0, 4])
print(distance_run_day1 + distance_run_day2)

[15 30 24 11  8]


We can also multiply, divide or square a numpy array and that operation will be applied to each element.

**Task:** You are given the heights and masses of a group of people in **lists**. The heights are in inches and the weights are in pounds (lbs).


1.   Convert the lists into numpy arrays and save these to new variable names.
2.   Look up how to convert from inches to metres and turn your numpy array for the heights into metres.
3.   Look up how to convert from pounds to kilograms and convert the values of your numpy array for the masses into kilograms.
4.   Calculate the BMI of all the people. Use the formula BMI = mass/(height^2). Save this new array to a variable.
5.   Look up the ranges of the BMI scale. Which of the people in our cohort are in a healthy weight range according to the NHS?


heights = [66.9, 71, 87, 61.8]

weights = [128, 165, 220, 123]



In [ ]:
# your answer
# heights is recorded in inches
# weights is recorded in pounds
heights = [66.9, 71, 87, 61.8]
weights = [128, 165, 220, 123]



## Introduction to Pandas

In this section, we will be looking to use an essential data science library: Pandas. Recall that Python libraries contain pre-written functions that we can use without having to rewrite them from scratch.

Pandas is a Python library used to handle data sets as well as perform analysis.

We start off by importing the two libraries.

In [ ]:
import pandas as pd

We first start off by bringing in the dataset. This dataset is an existing one that has been created and saved in another library called scikit-learn. A quick guide to the important variables:     



*   **data** is a pandas dataFrame of our data. It is essentially a table with column names given by the list **feature_names**.
*   **target** is a numpy array that tells us whether or not each tumour is cancerous. 0 means the tumour is cancerous and 1 means it is benign (non-cancerous). Each number label in target corresponds to a a classification of a row in our pandas dataframe.



In [ ]:
from sklearn import datasets
x = datasets.load_breast_cancer()
feature_names = list(x.feature_names)
data = pd.DataFrame(x.data, columns=feature_names)
target = x.target


It's always important to understand the syntax and semantics of the data we are dealing with. To best do this, we need a glipse of how the rows and columns are structured, and some actual entries in the data frame.

To see what the first few rows of the dataset look like, we can use the command **data.head()**. By default this prints out the first five rows. If you wish to print out more rows, we specify a number in the brackets.

The command **data.tail()** works the same way, except it prints the last few rows.

In [ ]:
data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [ ]:
data.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

In [ ]:
data.tail()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
564                0.2216          0.2060                  0.07115       0  
565                0.1628          0.2572                  0.06637       0  
566                0.1418          0.2218                  0.07820       0  
567                0.2650          0.4087                  0.12400       0  
568                0.0000          0.2871                  0.07039       1  

[5 rows x 31 columns]

**Accessing and adding columns to a dataframe**

Acessing or adding columns in a dataframe is useful if we want to read information from a particular column, or update the dataframe without having to copy over the existing data.

In order to access a column, we simply do:  

In [ ]:
data["mean radius"]

0      17.99
1      20.57
2      19.69
3      11.42
4      20.29
       ...  
564    21.56
565    20.13
566    16.60
567    20.60
568     7.76
Name: mean radius, Length: 569, dtype: float64

However, this returns what is called a Pandas series and is not what we typically want to work with (because of compatability reasons). We can turn automatically turn this Pandas series into a NumPy array (which we know how to work with!) by doing:

In [ ]:
data["mean radius"].values

array([17.99 , 20.57 , 19.69 , 11.42 , 20.29 , 12.45 , 18.25 , 13.71 ,
       13.   , 12.46 , 16.02 , 15.78 , 19.17 , 15.85 , 13.73 , 14.54 ,
       14.68 , 16.13 , 19.81 , 13.54 , 13.08 ,  9.504, 15.34 , 21.16 ,
       16.65 , 17.14 , 14.58 , 18.61 , 15.3  , 17.57 , 18.63 , 11.84 ,
       17.02 , 19.27 , 16.13 , 16.74 , 14.25 , 13.03 , 14.99 , 13.48 ,
       13.44 , 10.95 , 19.07 , 13.28 , 13.17 , 18.65 ,  8.196, 13.17 ,
       12.05 , 13.49 , 11.76 , 13.64 , 11.94 , 18.22 , 15.1  , 11.52 ,
       19.21 , 14.71 , 13.05 ,  8.618, 10.17 ,  8.598, 14.25 ,  9.173,
       12.68 , 14.78 ,  9.465, 11.31 ,  9.029, 12.78 , 18.94 ,  8.888,
       17.2  , 13.8  , 12.31 , 16.07 , 13.53 , 18.05 , 20.18 , 12.86 ,
       11.45 , 13.34 , 25.22 , 19.1  , 12.   , 18.46 , 14.48 , 19.02 ,
       12.36 , 14.64 , 14.62 , 15.37 , 13.27 , 13.45 , 15.06 , 20.26 ,
       12.18 ,  9.787, 11.6  , 14.42 , 13.61 ,  6.981, 12.18 ,  9.876,
       10.49 , 13.11 , 11.64 , 12.36 , 22.27 , 11.34 ,  9.777, 12.63 ,
      

We can even do this with several variables:

In [ ]:
data[["mean radius", "mean area"]].values

array([[  17.99, 1001.  ],
       [  20.57, 1326.  ],
       [  19.69, 1203.  ],
       ...,
       [  16.6 ,  858.1 ],
       [  20.6 , 1265.  ],
       [   7.76,  181.  ]])

If we now wish to add a column with values, such as our diagnosis for each tumour (which we saved to the target variable), we do the following:

In [ ]:
data["target"] = target

Note: "target" is a string that is passed in to name the column, and could therefore be anything!

**Task:** Take a look at **data** and familiarise yourself with the variables. Discuss with the person next to you and make sure you understand what all of them are!

In [ ]:
# your answer
# tumour values and descriptions

your answer: write a few points to note down what you explored

1.
2.
3.

**Task**: Write a function that retrieves the mean perimeter data, and divides all the values inside the mean perimeter column by 10 and return the new array. Print the result of the function at the end.

**Hint**: Remember how we manipulate Numpy arrays!

In [ ]:
# your answer
import pandas as pd
import numpy as np

def get_mean_perimeter_divided_by_ten(dataframe):

  mean_perimeter_array = dataframe["mean perimeter"].values
  divided_array = mean_perimeter_array / 10
  return divided_array


result_array = get_mean_perimeter_divided_by_ten(data)
print(result_array)

[12.28  13.29  13.     7.758 13.51   8.257 11.96   9.02   8.75   8.397
 10.27  10.36  13.24  10.37   9.36   9.673  9.474 10.81  13.     8.746
  8.563  6.034 10.25  13.72  11.    11.6    9.741 12.21  10.24  11.5
 12.48   7.793 11.28  12.79  10.7   11.01   9.363  8.261  9.554  8.84
  8.618  7.19  12.83   8.732  8.542 12.37   5.171  8.598  7.804  8.691
  7.472  8.721  7.571 12.03   9.726  7.334 12.55   9.555  8.261  5.434
  6.455  5.466  9.642  5.92   8.269  9.74   6.011  7.18   5.879  8.137
 12.36   5.879 11.42   9.043  7.919 10.41   8.791 12.02  14.37   8.319
  7.381  8.649 17.15  12.91   7.695 12.11   9.425 12.2    7.978  9.577
  9.457 10.02   8.474  8.66  10.03  13.24   7.779  6.211  7.434  9.448
  8.805  4.379  7.722  6.395  6.741  8.721  7.517  7.901 15.28   7.248
  6.25   8.215  9.783  6.864  5.584  7.653  5.874  9.864 10.57  11.42
  7.334 12.14  16.62   9.428  8.61   8.844  8.776 12.34   9.958 13.04
  7.908 10.17  10.62  10.2   12.02   8.172  7.472  7.306  9.685  7.3
  6.124 10.51

In [ ]:
#Teacher's answer:
def mean_perimeter_data(x):
  return x.values/10

**Multiple column accessing and indexing**

Although useful, we are not limited to indexing on column at a time. For example, we might also want to see how radius and area of a tumour relate so we want to analyse 2 variables at once, and because we are accessing more than 1 column, we have 2 sets of square brackets.


In [ ]:
data[["mean radius","mean area"]].values

array([[  17.99, 1001.  ],
       [  20.57, 1326.  ],
       [  19.69, 1203.  ],
       ...,
       [  16.6 ,  858.1 ],
       [  20.6 , 1265.  ],
       [   7.76,  181.  ]])

In [ ]:
data["target"] = target
data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [ ]:
print(target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

We are now going to take a look at some of the other functionalities that NumPy has to offer.

In particular, there are many pre-written functions that can calculate quantities of interest for us such as the mean, the median as well as the maximum and minimum values of a dataset.

To demonstrate this, we are now going to look at a breast cancer dataset. We are going to see if we can gain some insight into what distinguishes malign tumours from benign ones.We have 569 samples of which 212 are malign and 357 are benign.

We start off by examining a few variables of interest: tumour radius, perimeter, area and texture. We have created separate numpy arrays for each of these variables for the benign and malign tumours.

In [ ]:
data["target"] = x.target
patients_with_benign_tumour = data[data.target == 1]
patients_with_malign_tumour = data[data.target == 0]

benign_tumour_radii = patients_with_benign_tumour["mean radius"].values
malign_tumour_radii = patients_with_malign_tumour["mean radius"].values

benign_tumour_perimeter = patients_with_benign_tumour["mean perimeter"].values
malign_tumour_perimeter = patients_with_malign_tumour["mean perimeter"].values

benign_tumour_area = patients_with_benign_tumour["mean area"].values
malign_tumour_area = patients_with_malign_tumour["mean area"].values

benign_tumour_texture = patients_with_benign_tumour["mean texture"].values
malign_tumour_texture = patients_with_malign_tumour["mean texture"].values

We can use numpy functions to calculate basic statistics about benign and malign tumours. For example, we can do the following:

In [ ]:
mean_benign_radius = np.mean(benign_tumour_radii)

print(mean_benign_radius)

12.146523809523808


Doing this saves and outputs the mean radius of the benign tumours.
There are other numpy functions we can use:

np.max - provides the maximum value of the dataset
np.min - provides the minimum value of the dataset
np.std - provides the standard deviation of the dataset. The standard deviation is a measure of how "spread out" your data is around your mean value.

**Task**: Calculate the mean, maximum, minimum and standard deviation of the radii, perimeter, area and texture for both the benign and malign tumours. Can you say anything about how these values compare for the two types of tumours?

In [ ]:
# your answer


Useful to note, there is a Pandas-specific function that will provide you with a summary of your dataset is the **describe** function. For example:  

In [ ]:
data.describe()

## Introduction to Matplotlib

Matplotlib is a library for visualising data in Python. Using Matplotlib we can plot a wide variety of graphs to help us visualise our data. We first start off by **importing** a specific part of matplotlib (pyplot) into our **namespace**.

In [ ]:
import matplotlib.pyplot as plt

There are a variety of plots that can be done in matplotlib. We will only consider two in this session: scatter/line plots and histogram. Scatter plots can be created using the plt.scatter() function, e.g.:      

x = [1, 2, 3, 4]

y = [2, 4, 6, 8]

plt.scatter(x, y)

You can label your x and y axes by using the plt.xlabel() and plt.ylabel() functions. Simply make sure to provide the text you want as the label inside the brackets within quotation marks.

plt.xlabel("x")

plt.ylabel("y")

In [ ]:
x = [1, 2, 3, 4]
y = [2, 4, 6, 8]
plt.scatter(x, y)
plt.xlabel("x")
plt.ylabel("y")
plt.title("This is a scatter plot")
plt.show()

You can even plot multiple scatter plots on the same graph! In order to be able to distinguish them, you can add an extra argument inside the brackets called "label", which allows you to provide a name for each scatter plot. You then need to add plt.legend() afterwards to create a legend that will give you a colour code!

In [ ]:
x = [1, 2, 3, 4]
y1 = [2, 4, 6, 8]
y2 = [3, 6, 9, 12]
plt.scatter(x, y1, label = "A")
plt.scatter(x, y2, label = "B")
plt.xlabel("x")
plt.ylabel("y")
plt.legend()
plt.show()

We're now going to consider histograms which tell us how our data is distributed.

Below I generate 1000 random values which I assign to the variable **x**.

In [ ]:
mu, sigma = 20, 3
x = mu + sigma * np.random.randn(1000)
plt.hist(x, density=True, histtype = "step")
plt.xlabel("y")
plt.ylabel("Density")
plt.show()

**Task:** Using the numpy arrays for tumour radii, perimeters, areas and texture, plot histograms of these distributions for benign and malign tumours.

You will need to generate one histogram for each variable (radius, perimeter, area and texture). On each histogram, you will need to plot a histogram for the distribution of the variable for benign tumours as well as a histogram for the distribution of malign tumours. Be sure to include axes labels as well as a legend.

To create separate histograms for different variables, you will need to take a look at the plt.figure() method. The documentation can be found at [this link.](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.figure.html)

You might also need to take a look at [an example of using figure() with multiple plots.](https://matplotlib.org/stable/gallery/subplots_axes_and_figures/multiple_figs_demo.html#sphx-glr-gallery-subplots-axes-and-figures-multiple-figs-demo-py)


For each variable, can you determine a "cut-off" point where for a given value of the variable, say area, a tumour is more likely to be malign than benign?

In [ ]:
# your answer


**Task:** Create a scatter plot of mean radii and mean smoothness, where benign cases are presented in one colour, and malign cases are presented in another colour

In [ ]:
# your answer


## Introduction to documentation

Understanding and using documentation is an essential skill for any  programmer.

Imagine an book for kitchen objects. In this book, you have a picture of every appliance in your kitchen along with the user manual which contains how to use the device, including how to turn it on, off, what you can put inside of it and example recipes or applications.

Similar to this kitchen appliance manual, each tool in computer science (programming languages, libraries, datasets) has accompanying documentation. The documentation tells you why the tool was created in the first place (a.k.a. the problem the authors are trying to solve), how and where to use the tool in your applications and programs.

Here are documentation links for some of the libraries and languages we have used so far. Feel free to have a look now!


[Link to Python Documentation](https://www.python.org/)

[Link to Pandas Documentation](https://pandas.pydata.org/)

[Link to Numpy Documentation](https://numpy.org/doc/stable/)

[Link to Matplotlib Documentation](https://matplotlib.org/stable/index.html)
